src/Core/Services/MetadataProviders/SqlMetadataProvider.cs

```diff
@@ -1666,52 +1666,101 @@ IEnumerable<List<ForeignKeyDefinition>> foreignKeysForAllTargetEntities
-                                }
+        /// <summary>
+        /// Loops over all the foreign key definitions defined for the target entity in the source entity's definition
+        /// and adds to the set of validated FK definitions:
+        /// 1. All the FK definitions which actually map to a foreign key constraint defined in the database.
+        /// In such a case, if the source/target fields are also provided in the config, they are given precedence over the FK constraint.
+        /// 2. FK definitions for custom relationships defined by the user in the configuration file where no FK constraint exists between
+        /// the pair of (source, target) entities.
+        /// </summary>
+        /// <param name="fKDefinitionsToTarget">List of FK definitions defined from source to target.</param>
+        /// <returns>List of validated FK definitions from source to target.</returns>
+        private List<ForeignKeyDefinition> GetValidatedFKs(
+            List<ForeignKeyDefinition> fKDefinitionsToTarget)
+        {
+            List<ForeignKeyDefinition> validatedFKDefinitionsToTarget = new();
+            foreach (ForeignKeyDefinition fKDefinitionToTarget in fKDefinitionsToTarget)
+            {
+                // This code block adds FK definitions between source and target entities when there is an FK constraint defined
+                // in the database, either from source->target or target->source entities or both.
 
-                                // Only add the referenced columns if they have not been
-                                // specified in the configuration file.
-                                if (fk.ReferencedColumns.Count == 0)
-                                {
-                                    fk.ReferencedColumns.AddRange(inferredDefinition.ReferencedColumns);
-                                }
-                            }
-                        }
+                // Add the referencing and referenced columns for this foreign key definition for the target.
+                if (PairToFkDefinition is not null &&
+                    PairToFkDefinition.TryGetValue(fKDefinitionToTarget.Pair, out ForeignKeyDefinition? inferredFKDefinition))
+                {
+                    // Being here indicates that we inferred an FK constraint for the current FK.
+                    // The count of referencing and referenced columns being > 0 indicates that source.fields and target.fields 
+                    // have been specified in the config file. In this scenario, higher precedence is given to the fields configured through the config file. So, the existing FK definition is retained as is.   
+                    if (fKDefinitionToTarget.ReferencingColumns.Count > 0 && fKDefinitionToTarget.ReferencedColumns.Count > 0)
+                    {
+                        validatedFKDefinitionsToTarget.Add(fKDefinitionToTarget);
```

Comment:
if precedence is always going to be given to the config file fks, why do we need to do a lookup of inferredFKdefinition on line 1711. Why cant we directly add this fkDefinitionToTarget and skip the loop. 
Only if we dont find anything in the config is the need to lookup the inferred FK defintion or its inverse.

Reply:
Consider declaring a relationship in the config where entity: `Book`, relationship: `publishers` and source/target fields are provided by the user:

<img width="233" alt="image" src="https://github.com/Azure/data-api-builder/assets/34566234/e669a540-a5c5-4a86-abb7-ec75024b2989">,

For such a relationship where right cardinality is 1, we add an FK definition from both direction. In one of them we keep `Book` as the `referencing` table, and `Publisher` in the other.
At the end of the `FillInferredFKInfo()` however, we only want validated FK definitions. And this would be the FK definition where Book is the referencing entity.
If we do what you suggest, we would end up adding the invalid FK definition as well.

Reply:
> if precedence is always going to be given to the config file fks
And this behavior will change with support for `multiple-create`. I'll have to build on top of this PR.